File: backend/services/scoring/comps.js
/**
 * comps.js — Comp-anchored OW prediction.
 *
 * Finds real similar films, derives their actual opening-weekend figures,
 * and weights them by similarity to anchor the prediction.
 *
 * Two-pass enrichment strategy:
 *   Pass 1: score candidates on genre + market + recency (no API calls)
 *   Pass 2: enrich top-10 with getFilmDetails (revenue + budget + franchise)
 *           then re-score with full data, take best 5
 */
import db from '../../db/database.js';
import { getFilmDetails, getReleasedFilmsForComps } from '../tmdb.js';

/* ------------------------------------------------------------------ */
/* Constants                                                          */
/* ------------------------------------------------------------------ */

const GENRE_ID_MAP = {
  'Action': 28, 'Adventure': 12, 'Animation': 16, 'Comedy': 35,
  'Crime': 80, 'Documentary': 99, 'Drama': 18, 'Family': 10751,
  'Fantasy': 14, 'History': 36, 'Horror': 27, 'Music': 10402,
  'Mystery': 9648, 'Romance': 10749, 'Science Fiction': 878,
  'Thriller': 53, 'War': 10752, 'Western': 37,
};

// Origin-market splits: domestic fraction of WW revenue × domestic OW rate
const MARKET_OW_SPLITS = {
  hollywood: { domestic_pct: 0.40, ow_rate: 0.18 },
  bollywood: { domestic_pct: 0.65, ow_rate: 0.40 },
  korean:    { domestic_pct: 0.40, ow_rate: 0.35 },
  japanese:  { domestic_pct: 0.85, ow_rate: 0.45 },
  european:  { domestic_pct: 0.30, ow_rate: 0.12 },
};

const GLOBAL_OW_RATE     = 0.31;   // OW as % of worldwide total
const COMP_SCORE_CEILING = 300;    // $300M OW → comp_score 1.0 (in millions)
const MAX_ENRICH         = 10;     // API enrichment budget per call
const TOP_N              = 5;      // final comp set size
const WEIGHT_COMPS       = 3;      // use top-N for weighted average

/* ------------------------------------------------------------------ */
/* Shared helpers                                                     */
/* ------------------------------------------------------------------ */

function safeParse(v, fb) {
  if (v == null) return fb;
  if (typeof v === 'object') return v;
  try { return JSON.parse(v); } catch { return fb; }
}

function r2(n) { return Math.round(n * 100) / 100; }
function r4(n) { return Math.round(n * 10_000) / 10_000; }

function getBudgetTier(budget) {
  if (!budget || budget <= 0) return 'unknown';
  if (budget >= 150_000_000) return 'tentpole';
  if (budget >= 80_000_000)  return 'major';
  if (budget >= 25_000_000)  return 'mid';
  if (budget >= 5_000_000)   return 'indie';
  return 'micro';
}

/* ------------------------------------------------------------------ */
/* Candidate normalisation                                            */
/* ------------------------------------------------------------------ */

function normaliseDbRow(row) {
  const genreNames  = safeParse(row.genres, []);
  const collection  = safeParse(row.belongs_to_collection, null);
  return {
    db_id:                row.id,
    tmdb_id:              row.tmdb_id,
    title:                row.title,
    release_date:         row.release_date ?? null,
    revenue:              row.revenue       ?? 0,
    budget:               row.budget        ?? 0,
    budget_inferred:      !!row.budget_inferred,
    genre_ids:            genreNames.map((g) => GENRE_ID_MAP[g]).filter(Boolean),
    genre_names:          genreNames,
    market:               row.market        ?? 'hollywood',
    belongs_to_collection: collection,
    mpaa_rating:          row.mpaa_rating   ?? null,
    source:               'db',
  };
}

function normaliseDiscover(result, market) {
  return {
    db_id:                null,
    tmdb_id:              result.tmdb_id,
    title:                result.title,
    release_date:         result.release_date ?? null,
    revenue:              0,
    budget:               0,
    budget_inferred:      false,
    genre_ids:            result.genre_ids   ?? [],
    genre_names:          [],
    market,
    belongs_to_collection: null,
    mpaa_rating:          null,
    source:               'discover',
  };
}

/* ------------------------------------------------------------------ */
/* Revenue enrichment — upsert comp film into films table             */
/* ------------------------------------------------------------------ */

async function enrichCandidate(candidate) {
  // Already have revenue from DB? Use as-is.
  if (candidate.source === 'db' && candidate.revenue > 0) return candidate;

  // Check DB first (may have been upserted by a previous run).
  const existing = db.prepare('SELECT * FROM films WHERE tmdb_id = ?').get(candidate.tmdb_id);
  if (existing && (existing.revenue ?? 0) > 0) return normaliseDbRow(existing);

  // Fetch full details from TMDB.
  const d = await getFilmDetails(candidate.tmdb_id);
  if (!d) return null;
  if (!d.revenue || d.revenue === 0) return null; // no revenue → skip

  const genreNames = d.genres ?? [];
  const collection = d.belongs_to_collection;

  db.prepare(`
    INSERT INTO films (
      tmdb_id, title, overview, market, status, release_date,
      budget, revenue, runtime, genres, cast_top5, director,
      production_companies, production_countries, original_language,
      poster_path, trailer_url, belongs_to_collection, mpaa_rating,
      vote_average, vote_count, popularity, last_tmdb_sync
    ) VALUES (
      @tmdb_id, @title, @overview, @market, @status, @release_date,
      @budget, @revenue, @runtime, @genres, @cast_top5, @director,
      @production_companies, @production_countries, @original_language,
      @poster_path, @trailer_url, @belongs_to_collection, @mpaa_rating,
      @vote_average, @vote_count, @popularity, @last_tmdb_sync
    )
    ON CONFLICT(tmdb_id) DO UPDATE SET
      title                = excluded.title,
      market               = excluded.market,
      status               = excluded.status,
      budget               = excluded.budget,
      revenue              = excluded.revenue,
      genres               = excluded.genres,
      belongs_to_collection= excluded.belongs_to_collection,
      mpaa_rating          = excluded.mpaa_rating,
      vote_average         = excluded.vote_average,
      vote_count           = excluded.vote_count,
      popularity           = excluded.popularity,
      last_tmdb_sync       = excluded.last_tmdb_sync
  `).run({
    tmdb_id:               d.tmdb_id,
    title:                 d.title,
    overview:              d.overview ?? null,
    market:                d.market ?? candidate.market,
    status:                d.status ?? 'released',
    release_date:          d.release_date ?? null,
    budget:                d.budget  ?? 0,
    revenue:               d.revenue ?? 0,
    runtime:               d.runtime ?? null,
    genres:                JSON.stringify(genreNames),
    cast_top5:             JSON.stringify(d.cast_top5 ?? []),
    director:              JSON.stringify(d.director  ?? null),
    production_companies:  JSON.stringify(d.production_companies ?? []),
    production_countries:  JSON.stringify(d.production_countries ?? []),
    original_language:     d.original_language ?? null,
    poster_path:           d.poster_path ?? null,
    trailer_url:           d.trailer_url ?? null,
    belongs_to_collection: collection ? JSON.stringify(collection) : null,
    mpaa_rating:           d.mpaa_rating ?? null,
    vote_average:          d.vote_average ?? 0,
    vote_count:            d.vote_count   ?? 0,
    popularity:            d.popularity   ?? 0,
    last_tmdb_sync:        new Date().toISOString(),
  });

  const row = db.prepare('SELECT * FROM films WHERE tmdb_id = ?').get(d.tmdb_id);
  return row ? normaliseDbRow(row) : null;
}

/* ------------------------------------------------------------------ */
/* Similarity scoring                                                 */
/* ------------------------------------------------------------------ */

// Lightweight first-pass: genre + market + recency only (no API needed).
function prelimScore(film, candidate) {
  const filmGenreNames = safeParse(film.genres, []);
  const filmPrimaryId  = filmGenreNames.map((g) => GENRE_ID_MAP[g]).find(Boolean);
  let pts = 0;
  if (filmPrimaryId && candidate.genre_ids[0] === filmPrimaryId) pts += 3;
  if ((film.market ?? 'hollywood') === candidate.market)          pts += 3;
  const yr = candidate.release_date ? parseInt(candidate.release_date.slice(0, 4), 10) : null;
  if (yr) {
    const ago = new Date().getFullYear() - yr;
    if (ago <= 2) pts += 2;
    else if (ago <= 5) pts += 1;
  }
  return pts;
}

// Full similarity after enrichment (13-point system).
function fullScore(film, candidate) {
  const filmGenreNames  = safeParse(film.genres, []);
  const filmPrimaryId   = filmGenreNames.map((g) => GENRE_ID_MAP[g]).find(Boolean);
  const filmCollection  = safeParse(film.belongs_to_collection, null);
  const filmHasColl     = !!filmCollection;
  const filmBudget      = film.budget_inferred ? 150_000_000 : (film.budget ?? 0);
  const filmTier        = getBudgetTier(filmBudget);
  const filmMpaa        = film.mpaa_rating ?? null;

  const candBudget      = candidate.budget_inferred ? 150_000_000 : (candidate.budget ?? 0);
  const candTier        = getBudgetTier(candBudget);
  const candHasColl     = !!candidate.belongs_to_collection;
  const candMpaa        = candidate.mpaa_rating ?? null;
  const candPrimaryId   = candidate.genre_ids[0];

  let pts = 0;
  const reasons = [];

  if (filmPrimaryId && candPrimaryId && filmPrimaryId === candPrimaryId) {
    pts += 3;
    reasons.push(`Same genre (${filmGenreNames[0] ?? 'Action'})`);
  }
  if ((film.market ?? 'hollywood') === candidate.market) {
    pts += 3;
    const label = (candidate.market.charAt(0).toUpperCase() + candidate.market.slice(1));
    reasons.push(`Same market (${label})`);
  }
  if (filmTier === candTier) {
    pts += 2;
    reasons.push(`Same budget tier (${filmTier})`);
  }
  if (filmHasColl === candHasColl) {
    pts += 2;
    reasons.push('Same franchise structure');
  }
  if (filmMpaa && candMpaa && filmMpaa === candMpaa) {
    pts += 1;
    reasons.push(`Same MPAA rating (${filmMpaa})`);
  }
  const yr = candidate.release_date ? parseInt(candidate.release_date.slice(0, 4), 10) : null;
  if (yr) {
    const ago = new Date().getFullYear() - yr;
    if (ago <= 2) { pts += 2; reasons.push(`Recent release (${yr})`); }
    else if (ago <= 5) { pts += 1; reasons.push(`Recent release (${yr})`); }
  }

  return { similarity_score: r4(pts / 13), match_reasons: reasons };
}

/* ------------------------------------------------------------------ */
/* OW estimate from comp revenue                                      */
/* ------------------------------------------------------------------ */

function compOW(revenue, market) {
  const sp = MARKET_OW_SPLITS[market] ?? MARKET_OW_SPLITS.hollywood;
  return {
    ow_origin: r2((revenue * sp.domestic_pct * sp.ow_rate) / 1_000_000),
    ow_global: r2((revenue * GLOBAL_OW_RATE) / 1_000_000),
  };
}

/* ------------------------------------------------------------------ */
/* Cache path                                                         */
/* ------------------------------------------------------------------ */

function calcFromCached(rows) {
  const top3 = rows.slice(0, WEIGHT_COMPS);
  const tw   = top3.reduce((s, r) => s + r.similarity_score, 0);
  const wOri = top3.reduce((s, r) => s + r.comp_actual_ow_origin * r.similarity_score, 0) / tw;
  const wGlo = top3.reduce((s, r) => s + r.comp_actual_ow_global * r.similarity_score, 0) / tw;
  return {
    score:                  r4(Math.min(wGlo / COMP_SCORE_CEILING, 1.0)),
    weighted_ow_origin_usd: r2(wOri),
    weighted_ow_global_usd: r2(wGlo),
    comps_found:            rows.length,
    comps_used:             rows.map((r) => ({
      title:               r.title,
      release_date:        r.release_date,
      similarity_score:    r.similarity_score,
      match_reasons:       safeParse(r.match_reasons, []),
      actual_ow_origin_usd: r.comp_actual_ow_origin,
      actual_ow_global_usd: r.comp_actual_ow_global,
    })),
    data_available: true,
  };
}

/* ================================================================== */
/* computeCompScore(film)                                             */
/* ================================================================== */

export async function computeCompScore(film) {
  const filmDbId = film.id ?? null;

  /* ── Step 1: cache check ────────────────────────────────────────── */
  if (filmDbId) {
    const cached = db.prepare(`
      SELECT ca.*, f.title, f.revenue, f.release_date, f.market
      FROM comp_anchors ca
      JOIN films f ON ca.comp_film_id = f.id
      WHERE ca.upcoming_film_id = ?
      ORDER BY ca.similarity_score DESC
      LIMIT ?
    `).all(filmDbId, TOP_N);

    if (cached.length >= 3 && cached.every((r) => r.comp_actual_ow_global > 0)) {
      console.log(`[COMPS] Using cached comps for "${film.title}"`);
      return calcFromCached(cached);
    }
  }

  /* ── Step 2: find candidates ────────────────────────────────────── */
  const filmGenreNames = safeParse(film.genres, []);
  const primaryGenreName = filmGenreNames[0] ?? 'Action';
  const primaryGenreId   = GENRE_ID_MAP[primaryGenreName] ?? 28;
  const filmMarket       = film.market ?? 'hollywood';
  const filmBudget       = film.budget_inferred ? 150_000_000 : (film.budget ?? 0);

  console.log(`[COMPS] Finding comps for "${film.title}" (${filmMarket}, genre=${primaryGenreName})`);

  // TMDB discover candidates
  const discoverResults = await getReleasedFilmsForComps(
    [primaryGenreId],
    filmMarket,
    filmBudget || 150_000_000
  );

  // DB candidates (already have revenue)
  const dbRows = db.prepare(`
    SELECT * FROM films
    WHERE status = 'released'
      AND market = ?
      AND revenue > 0
      AND tmdb_id != ?
  `).all(filmMarket, film.tmdb_id ?? 0);

  // Merge, deduplicate by tmdb_id
  const seen       = new Set();
  const candidates = [];

  for (const row of dbRows) {
    seen.add(row.tmdb_id);
    candidates.push(normaliseDbRow(row));
  }
  for (const r of (discoverResults ?? [])) {
    if (!seen.has(r.tmdb_id)) {
      seen.add(r.tmdb_id);
      // If the DB already has this film (without revenue) use its data
      const existing = db.prepare('SELECT * FROM films WHERE tmdb_id = ?').get(r.tmdb_id);
      candidates.push(existing ? normaliseDbRow(existing) : normaliseDiscover(r, filmMarket));
    }
  }

  console.log(`[COMPS] ${candidates.length} candidates found`);

  /* ── Step 3: prelim score → enrich top MAX_ENRICH ──────────────── */
  const byPrelim = candidates
    .map((c) => ({ ...c, prelim: prelimScore(film, c) }))
    .sort((a, b) => b.prelim - a.prelim)
    .slice(0, MAX_ENRICH);

  const enriched = [];
  for (const candidate of byPrelim) {
    const result = await enrichCandidate(candidate);
    if (result && result.revenue > 0) enriched.push(result);
  }

  console.log(`[COMPS] ${enriched.length} candidates after revenue enrichment`);

  if (enriched.length === 0) {
    return {
      score: 0.30,
      weighted_ow_origin_usd: null,
      weighted_ow_global_usd: null,
      comps_found: 0,
      comps_used:  [],
      data_available: false,
      reason: 'No comparable films found',
    };
  }

  /* ── Full similarity + OW estimation ────────────────────────────── */
  const scored = enriched.map((c) => {
    const { similarity_score, match_reasons } = fullScore(film, c);
    const { ow_origin, ow_global } = compOW(c.revenue, c.market);
    return { ...c, similarity_score, match_reasons, ow_origin, ow_global };
  }).sort((a, b) => b.similarity_score - a.similarity_score);

  const top5 = scored.slice(0, TOP_N);

  /* ── Step 4: persist comp_anchors ───────────────────────────────── */
  if (filmDbId) {
    db.prepare('DELETE FROM comp_anchors WHERE upcoming_film_id = ?').run(filmDbId);
    const ins = db.prepare(`
      INSERT INTO comp_anchors
        (upcoming_film_id, comp_film_id, similarity_score,
         match_reasons, comp_actual_ow_origin, comp_actual_ow_global)
      VALUES (?, ?, ?, ?, ?, ?)
    `);
    for (const c of top5) {
      if (c.db_id) {
        ins.run(filmDbId, c.db_id, c.similarity_score,
                JSON.stringify(c.match_reasons), c.ow_origin, c.ow_global);
      }
    }
  }

  /* ── Weighted average (top WEIGHT_COMPS) ────────────────────────── */
  const top3 = top5.slice(0, WEIGHT_COMPS);
  const tw   = top3.reduce((s, c) => s + c.similarity_score, 0);
  const wOri = top3.reduce((s, c) => s + c.ow_origin * c.similarity_score, 0) / tw;
  const wGlo = top3.reduce((s, c) => s + c.ow_global * c.similarity_score, 0) / tw;

  const score = r4(Math.min(wGlo / COMP_SCORE_CEILING, 1.0));

  console.log(`[COMPS] weighted OW global = $${r2(wGlo)}M → comp_score = ${score}`);

  return {
    score,
    weighted_ow_origin_usd: r2(wOri),
    weighted_ow_global_usd: r2(wGlo),
    comps_found: top5.length,
    comps_used:  top5.map((c) => ({
      title:                c.title,
      release_date:         c.release_date,
      similarity_score:     c.similarity_score,
      match_reasons:        c.match_reasons,
      actual_ow_origin_usd: c.ow_origin,
      actual_ow_global_usd: c.ow_global,
    })),
    data_available: true,
  };
}

File: backend/services/scoring/comps.py
import json
from datetime import datetime, timezone
from ...db.database import db, execute_write
from ..tmdb import get_film_details, get_released_films_for_comps

GENRE_ID_MAP = {
    'Action': 28, 'Adventure': 12, 'Animation': 16, 'Comedy': 35,
    'Crime': 80, 'Documentary': 99, 'Drama': 18, 'Family': 10751,
    'Fantasy': 14, 'History': 36, 'Horror': 27, 'Music': 10402,
    'Mystery': 9648, 'Romance': 10749, 'Science Fiction': 878,
    'Thriller': 53, 'War': 10752, 'Western': 37,
}

MARKET_OW_SPLITS = {
    'hollywood': {'domestic_pct': 0.40, 'ow_rate': 0.18},
    'bollywood': {'domestic_pct': 0.65, 'ow_rate': 0.40},
    'korean':    {'domestic_pct': 0.40, 'ow_rate': 0.35},
    'japanese':  {'domestic_pct': 0.85, 'ow_rate': 0.45},
    'european':  {'domestic_pct': 0.30, 'ow_rate': 0.12},
}

GLOBAL_OW_RATE     = 0.31
COMP_SCORE_CEILING = 300
MAX_ENRICH         = 10
TOP_N              = 5
WEIGHT_COMPS       = 3


def _safe_parse(v, fallback):
    if v is None:
        return fallback
    if isinstance(v, (dict, list)):
        return v
    try:
        return json.loads(v)
    except Exception:
        return fallback


def _r2(n): return round(n * 100) / 100
def _r4(n): return round(n * 10_000) / 10_000


def _budget_tier(budget):
    if not budget or budget <= 0: return 'unknown'
    if budget >= 150_000_000:     return 'tentpole'
    if budget >= 80_000_000:      return 'major'
    if budget >= 25_000_000:      return 'mid'
    if budget >= 5_000_000:       return 'indie'
    return 'micro'


def _normalise_db_row(row):
    row = dict(row)
    genre_names = _safe_parse(row.get('genres'), [])
    collection  = _safe_parse(row.get('belongs_to_collection'), None)
    return {
        'db_id':                row['id'],
        'tmdb_id':              row['tmdb_id'],
        'title':                row['title'],
        'release_date':         row.get('release_date'),
        'revenue':              row.get('revenue') or 0,
        'budget':               row.get('budget') or 0,
        'budget_inferred':      bool(row.get('budget_inferred')),
        'genre_ids':            [GENRE_ID_MAP[g] for g in genre_names if g in GENRE_ID_MAP],
        'genre_names':          genre_names,
        'market':               row.get('market') or 'hollywood',
        'belongs_to_collection': collection,
        'mpaa_rating':          row.get('mpaa_rating'),
        'source':               'db',
    }


def _normalise_discover(result, market):
    return {
        'db_id':                None,
        'tmdb_id':              result['tmdb_id'],
        'title':                result['title'],
        'release_date':         result.get('release_date'),
        'revenue':              0,
        'budget':               0,
        'budget_inferred':      False,
        'genre_ids':            result.get('genre_ids') or [],
        'genre_names':          [],
        'market':               market,
        'belongs_to_collection': None,
        'mpaa_rating':          None,
        'source':               'discover',
    }


async def _enrich_candidate(candidate):
    if candidate['source'] == 'db' and candidate['revenue'] > 0:
        return candidate

    existing = db.execute('SELECT * FROM films WHERE tmdb_id=?', (candidate['tmdb_id'],)).fetchone()
    if existing and (dict(existing).get('revenue') or 0) > 0:
        return _normalise_db_row(existing)

    d = await get_film_details(candidate['tmdb_id'])
    if not d or not d.get('revenue'):
        return None

    genre_names = d.get('genres') or []
    collection  = d.get('belongs_to_collection')

    execute_write(
        '''INSERT INTO films (
              tmdb_id, title, overview, market, status, release_date,
              budget, revenue, runtime, genres, cast_top5, director,
              production_companies, production_countries, original_language,
              poster_path, trailer_url, belongs_to_collection, mpaa_rating,
              vote_average, vote_count, popularity, last_tmdb_sync
           ) VALUES (
              :tmdb_id, :title, :overview, :market, :status, :release_date,
              :budget, :revenue, :runtime, :genres, :cast_top5, :director,
              :production_companies, :production_countries, :original_language,
              :poster_path, :trailer_url, :belongs_to_collection, :mpaa_rating,
              :vote_average, :vote_count, :popularity, :last_tmdb_sync
           )
           ON CONFLICT(tmdb_id) DO UPDATE SET
              title=excluded.title, market=excluded.market, status=excluded.status,
              budget=excluded.budget, revenue=excluded.revenue, genres=excluded.genres,
              belongs_to_collection=excluded.belongs_to_collection,
              mpaa_rating=excluded.mpaa_rating, vote_average=excluded.vote_average,
              vote_count=excluded.vote_count, popularity=excluded.popularity,
              last_tmdb_sync=excluded.last_tmdb_sync''',
        {
            'tmdb_id':               d['tmdb_id'],
            'title':                 d['title'],
            'overview':              d.get('overview'),
            'market':                d.get('market') or candidate['market'],
            'status':                d.get('status') or 'released',
            'release_date':          d.get('release_date'),
            'budget':                d.get('budget') or 0,
            'revenue':               d.get('revenue') or 0,
            'runtime':               d.get('runtime'),
            'genres':                json.dumps(genre_names),
            'cast_top5':             json.dumps(d.get('cast_top5') or []),
            'director':              json.dumps(d.get('director')),
            'production_companies':  json.dumps(d.get('production_companies') or []),
            'production_countries':  json.dumps(d.get('production_countries') or []),
            'original_language':     d.get('original_language'),
            'poster_path':           d.get('poster_path'),
            'trailer_url':           d.get('trailer_url'),
            'belongs_to_collection': json.dumps(collection) if collection else None,
            'mpaa_rating':           d.get('mpaa_rating'),
            'vote_average':          d.get('vote_average') or 0,
            'vote_count':            d.get('vote_count') or 0,
            'popularity':            d.get('popularity') or 0,
            'last_tmdb_sync':        datetime.now(timezone.utc).isoformat(),
        }
    )
    row = db.execute('SELECT * FROM films WHERE tmdb_id=?', (d['tmdb_id'],)).fetchone()
    return _normalise_db_row(row) if row else None


def _prelim_score(film, candidate):
    film_genre_names = _safe_parse(film.get('genres'), [])
    film_primary_id  = next((GENRE_ID_MAP[g] for g in film_genre_names if g in GENRE_ID_MAP), None)
    pts = 0
    if film_primary_id and candidate['genre_ids'] and candidate['genre_ids'][0] == film_primary_id:
        pts += 3
    if (film.get('market') or 'hollywood') == candidate['market']:
        pts += 3
    yr = int(candidate['release_date'][:4]) if candidate.get('release_date') else None
    if yr:
        ago = datetime.now(timezone.utc).year - yr
        if ago <= 2:   pts += 2
        elif ago <= 5: pts += 1
    return pts


def _full_score(film, candidate):
    film_genre_names = _safe_parse(film.get('genres'), [])
    film_primary_id  = next((GENRE_ID_MAP[g] for g in film_genre_names if g in GENRE_ID_MAP), None)
    film_collection  = _safe_parse(film.get('belongs_to_collection'), None)
    film_has_coll    = bool(film_collection)
    film_budget      = 150_000_000 if film.get('budget_inferred') else (film.get('budget') or 0)
    film_tier        = _budget_tier(film_budget)
    film_mpaa        = film.get('mpaa_rating')

    cand_budget = 150_000_000 if candidate['budget_inferred'] else (candidate['budget'] or 0)
    cand_tier   = _budget_tier(cand_budget)
    cand_has_coll = bool(candidate['belongs_to_collection'])
    cand_mpaa     = candidate['mpaa_rating']
    cand_primary  = candidate['genre_ids'][0] if candidate['genre_ids'] else None

    pts, reasons = 0, []
    if film_primary_id and cand_primary and film_primary_id == cand_primary:
        pts += 3
        reasons.append(f'Same genre ({film_genre_names[0] if film_genre_names else "Action"})')
    if (film.get('market') or 'hollywood') == candidate['market']:
        pts += 3
        label = candidate['market'].capitalize()
        reasons.append(f'Same market ({label})')
    if film_tier == cand_tier:
        pts += 2
        reasons.append(f'Same budget tier ({film_tier})')
    if film_has_coll == cand_has_coll:
        pts += 2
        reasons.append('Same franchise structure')
    if film_mpaa and cand_mpaa and film_mpaa == cand_mpaa:
        pts += 1
        reasons.append(f'Same MPAA rating ({film_mpaa})')
    yr = int(candidate['release_date'][:4]) if candidate.get('release_date') else None
    if yr:
        ago = datetime.now(timezone.utc).year - yr
        if ago <= 2:
            pts += 2; reasons.append(f'Recent release ({yr})')
        elif ago <= 5:
            pts += 1; reasons.append(f'Recent release ({yr})')

    return {'similarity_score': _r4(pts / 13), 'match_reasons': reasons}


def _comp_ow(revenue, market):
    sp = MARKET_OW_SPLITS.get(market) or MARKET_OW_SPLITS['hollywood']
    return {
        'ow_origin': _r2(revenue * sp['domestic_pct'] * sp['ow_rate'] / 1_000_000),
        'ow_global': _r2(revenue * GLOBAL_OW_RATE / 1_000_000),
    }


def _calc_from_cached(rows):
    top3 = rows[:WEIGHT_COMPS]
    tw   = sum(r['similarity_score'] for r in top3)
    w_ori = sum(r['comp_actual_ow_origin'] * r['similarity_score'] for r in top3) / tw
    w_glo = sum(r['comp_actual_ow_global'] * r['similarity_score'] for r in top3) / tw
    return {
        'score':                  _r4(min(w_glo / COMP_SCORE_CEILING, 1.0)),
        'weighted_ow_origin_usd': _r2(w_ori),
        'weighted_ow_global_usd': _r2(w_glo),
        'comps_found':            len(rows),
        'comps_used':             [
            {'title': r['title'], 'release_date': r.get('release_date'),
             'similarity_score': r['similarity_score'],
             'match_reasons': _safe_parse(r.get('match_reasons'), []),
             'actual_ow_origin_usd': r['comp_actual_ow_origin'],
             'actual_ow_global_usd': r['comp_actual_ow_global']}
            for r in rows
        ],
        'data_available': True,
    }


async def compute_comp_score(film):
    film_db_id = film.get('id')

    if film_db_id:
        cached = db.execute(
            '''SELECT ca.*, f.title, f.revenue, f.release_date, f.market
               FROM comp_anchors ca JOIN films f ON ca.comp_film_id=f.id
               WHERE ca.upcoming_film_id=?
               ORDER BY ca.similarity_score DESC LIMIT ?''',
            (film_db_id, TOP_N)
        ).fetchall()
        cached = [dict(r) for r in cached]
        if len(cached) >= 3 and all(r['comp_actual_ow_global'] > 0 for r in cached):
            print(f'[COMPS] Using cached comps for "{film.get("title")}"')
            return _calc_from_cached(cached)

    film_genre_names = _safe_parse(film.get('genres'), [])
    primary_genre_name = film_genre_names[0] if film_genre_names else 'Action'
    primary_genre_id   = GENRE_ID_MAP.get(primary_genre_name, 28)
    film_market        = film.get('market') or 'hollywood'
    film_budget        = 150_000_000 if film.get('budget_inferred') else (film.get('budget') or 0)

    print(f'[COMPS] Finding comps for "{film.get("title")}" ({film_market}, genre={primary_genre_name})')

    discover_results = await get_released_films_for_comps(
        [primary_genre_id], film_market, film_budget or 150_000_000
    )

    db_rows = db.execute(
        'SELECT * FROM films WHERE status=? AND market=? AND revenue>0 AND tmdb_id!=?',
        ('released', film_market, film.get('tmdb_id') or 0)
    ).fetchall()

    seen, candidates = set(), []
    for row in db_rows:
        row = dict(row)
        seen.add(row['tmdb_id'])
        candidates.append(_normalise_db_row(row))
    for r in (discover_results or []):
        if r['tmdb_id'] not in seen:
            seen.add(r['tmdb_id'])
            existing = db.execute('SELECT * FROM films WHERE tmdb_id=?', (r['tmdb_id'],)).fetchone()
            candidates.append(_normalise_db_row(existing) if existing else _normalise_discover(r, film_market))

    print(f'[COMPS] {len(candidates)} candidates found')

    by_prelim = sorted(
        [{'candidate': c, 'prelim': _prelim_score(film, c)} for c in candidates],
        key=lambda x: x['prelim'], reverse=True
    )[:MAX_ENRICH]

    enriched = []
    for item in by_prelim:
        result = await _enrich_candidate(item['candidate'])
        if result and result['revenue'] > 0:
            enriched.append(result)

    print(f'[COMPS] {len(enriched)} candidates after revenue enrichment')

    if not enriched:
        return {
            'score': 0.30, 'weighted_ow_origin_usd': None, 'weighted_ow_global_usd': None,
            'comps_found': 0, 'comps_used': [], 'data_available': False,
            'reason': 'No comparable films found',
        }

    scored = []
    for c in enriched:
        sim = _full_score(film, c)
        ow  = _comp_ow(c['revenue'], c['market'])
        scored.append({**c, **sim, **ow})
    scored.sort(key=lambda x: x['similarity_score'], reverse=True)
    top5 = scored[:TOP_N]

    if film_db_id:
        execute_write('DELETE FROM comp_anchors WHERE upcoming_film_id=?', (film_db_id,))
        for c in top5:
            if c.get('db_id'):
                execute_write(
                    '''INSERT INTO comp_anchors
                       (upcoming_film_id, comp_film_id, similarity_score,
                        match_reasons, comp_actual_ow_origin, comp_actual_ow_global)
                       VALUES (?,?,?,?,?,?)''',
                    (film_db_id, c['db_id'], c['similarity_score'],
                     json.dumps(c['match_reasons']), c['ow_origin'], c['ow_global'])
                )

    top3 = top5[:WEIGHT_COMPS]
    tw   = sum(c['similarity_score'] for c in top3)
    w_ori = sum(c['ow_origin'] * c['similarity_score'] for c in top3) / tw
    w_glo = sum(c['ow_global'] * c['similarity_score'] for c in top3) / tw
    score = _r4(min(w_glo / COMP_SCORE_CEILING, 1.0))

    print(f'[COMPS] weighted OW global = ${_r2(w_glo)}M → comp_score = {score}')

    return {
        'score':                  score,
        'weighted_ow_origin_usd': _r2(w_ori),
        'weighted_ow_global_usd': _r2(w_glo),
        'comps_found':            len(top5),
        'comps_used': [
            {'title': c['title'], 'release_date': c.get('release_date'),
             'similarity_score': c['similarity_score'], 'match_reasons': c['match_reasons'],
             'actual_ow_origin_usd': c['ow_origin'], 'actual_ow_global_usd': c['ow_global']}
            for c in top5
        ],
        'data_available': True,
    }
